2497 ESSignalProcessor for custom handling of ES document indexing

albertisfu opened the pull request:
This PR introduces some improvements  in the way Elasticsearch documents are updated when related objects are updated in the database, ensuring that the documents remain synchronized.

`django-elasticsearch-dsl` provides built-in signals that when used alongside `related_models`, allows documents stay in sync when related objects are updated in the database. However, there are problems with this approach. Firstly, any change in related documents, including changes on fields we're not indexing will trigger updates. Secondly, documents are completely updated, meaning all fields are sent to Elasticsearch, even if not all of them have changed. So this approach is not performance efficient.

The solution to this problem involves triggering document updates only when indexed fields on related documents change. Additionally, we should only perform partial updates in Elasticsearch, including only the fields that have changed.

To implement this solution, custom signals for each Django model involved in each Elasticsearch document mapping must be registered. However, this could lead to a lot of repeated code and maintenance will be complex.

So a more efficient strategy would be to handle all of these signals in a centralized class. So the `ESSignalProcessor` class has been introduced.

- In order to solve the problem related to only triggering document updates when indexed fields change on related documents, the [django-model-utils](https://django-model-utils.readthedocs.io/en/latest/index.html) [Field Tracker](https://django-model-utils.readthedocs.io/en/latest/utilities.html#field-tracker) is employed.

For the Field Tracker to function correctly, it must be added to each related model from which we wish to track field changes. For example:

```
es_pa_field_tracker = FieldTracker(
        fields=[
            "case_name",
            "citation_count",
            "date_filed",
            "slug",
            "docket_id",
            "judges",
            "nature_of_suit",
            "precedential_status",
        ]
    )
```

No migrations are required.

- In order to address the problem of sending the whole document to Elasticsearch each time for updates, [partial updates](https://elasticsearch-dsl.readthedocs.io/en/latest/api.html#elasticsearch_dsl.Document.update) are employed. As a result, only affected fields are sent to Elasticsearch for their respective updates.

`ESSignalProcessor` handles the following types of document indexing/updates.

- On models `save`:
  - Main document indexing/update. This action adds the complete document to the Elasticsearch index when a new instance of the main model is created or when an existing instance is updated.
  - Related model fields partial update. This action updates related model fields when indexed fields change in the related instance.
- On models `delete`:
  - Removal of the document from Elasticsearch when the main model instance is deleted. This signal is not triggered when related instances are deleted. As objects are related using `on_delete=models.CASCADE`, if a parent document is deleted, all child objects will also be deleted, and only the main model instances will trigger the removal from the Elasticsearch index.
- On m2m change:
  - Triggers partial updates when a `m2m` relation included in indexed fields is added or removed.
- On ForeignKey reverse relations changes:
  - Indexed fields that depend on reverse relations are updated when a reverse relation is added or removed.

The process to add new/update documents and models to the `ESSignalProcessor` is as follows:

- Keep track of fields mapped in the Elasticsearch document and their model relations.
- For direct ForeignKey relations, add a `FieldTracker` on each related model:
  -  e.g: `Parenthetical`: `es_pa_field_tracker = FieldTracker(fields=["score", "text"]) `
- Register a document-models and fields/actions mapping for handling signals.
- Add the document-models and fields/actions mapping to `ESSignalProcessor` instance, in order to register it in `signals.py`
  

The document-models and fields/actions mapping looks like this, with comments added to explain how it works.
```
pa_fields_mapping = {
    "save": {  # Key, describing the action on which signals will be registered, in this case "on save".
        Docket: { # The model on which the signal will be registered.
            "opinion__cluster__docket": { # The relation relative to the main model e.g: 'ParentheticalGroup', according to the ES document mapping.
                "docket_number": ("docketNumber", "raw"), # List of fields to keep in sync, field name in DB as key. Two tuple with the Field name in ES mapping and field value to register. 
                "court_id": ("court_id", "raw"), # "raw" is the field value stored in DB, "display" will use the human readable value.
            }
        },
        Opinion: {
            "opinion": {
                "author_id": ("author_id", "raw"),
                "cluster_id": ("cluster_id", "raw"),
                "extracted_by_ocr": ("opinion_extracted_by_ocr", "raw"),
            },
        },
        OpinionCluster: {
            "representative__describing_opinion__cluster": {
                "slug": ("describing_opinion_cluster_slug", "raw")
            },
            "opinion__cluster": {
                "case_name": ("caseName", "raw"),
                "citation_count": ("citeCount", "raw"),
                "date_filed": ("dateFiled", "raw"),
                "slug": ("opinion_cluster_slug", "raw"),
                "docket_id": ("docket_id", "raw"),
                "judges": ("judge", "raw"),
                "nature_of_suit": ("suitNature", "raw"),
                "precedential_status": (
                    "status",
                    "display",
                ),  # This should be returned as display()
            },
        },
        Parenthetical: {
            "representative": {
                "score": ("representative_score", "raw"),
                "text": ("representative_text", "raw"),
            },
        },
        ParentheticalGroup: {}, # For the main model, fields mapping is not required, since all its fields will be indexed/updated.
    },
    "delete": {ParentheticalGroup: {}}, # "delete" action, this only applies to the main model, no fields mapping is required.
    "m2m": { # "m2m" action.
        OpinionCluster.panel.through: { # Use the through model for the m2m relation.
            "opinion__cluster": { # The relation relative to the main model 'ParentheticalGroup', according to the ES document mapping.
                "panel_ids": ("panel_ids", "raw"),
            },
        },
        OpinionsCited: {
            "opinion": {
                "cites": ("cites", "raw"),
            },
        },
    },
    "reverse": {  # "reverse" action.
        Citation: {
            "opinion__cluster": { # The relation relative to the main model 'ParentheticalGroup', according to the ES document mapping.
                "all": ["citation"], # Reverse relation supports instance types, so each instance type can be indexed in different fields. Use all for all types.
                Citation.NEUTRAL: ["citation", "neutralCite"], # Set the instance type, and a list of ES fields to update.
                Citation.LEXIS: ["citation", "lexisCite"],
            },
        }
    },
}

```

Let me know what do you think.

mlissner commented on cl/settings/third_party/elasticsearch.py:
```diff
@@ -4,9 +4,10 @@
 
 ELASTICSEARCH_DISABLED = env(
```
Should we just remove this setting completely?

mlissner replied:
Never mind, I remembered this will morph into a setting for PA, and another for OA, and C, and so forth...

albertisfu replied:
Yes, exactly. On OA, I'll change the name of this setting so that we can reuse it to control the search engine transition.

mlissner commented on cl/lib/custom_es_signal_processor.py:
I'd rename this as just es_signal_processor.py. We know it's custom, and it'll make it more concise.

albertisfu replied:
Got it, changed.

mlissner commented on cl/search/signals.py:
```diff
@@ -0,0 +1,81 @@
+    },
+}
+
+pa_signal_processor = ESSignalProcessor
+pa_signal_processor(
+    (ParentheticalGroup, ParentheticalGroupDocument, pa_fields_mapping)
+)
```
Why not:

```suggestion
_pa_signal_processor = ESSignalProcessor(
    ParentheticalGroup, ParentheticalGroupDocument, pa_fields_mapping,
)
```

I think that's the same, except:

 - It doesn't pass a tuple, and instead passes arguments, making the invocation more obvious
 - It eliminates line 78, which I think you were doing just for self-documentation purposes
 - Is one line shorter?

albertisfu replied:
Yeah, this is clearer, I change it.

mlissner commented on cl/search/signals.py:
```diff
@@ -0,0 +1,81 @@
+    ParentheticalGroup,
+)
+
+pa_fields_mapping = {
```
```suggestion
pa_field_mapping = {
```

mlissner commented on cl/search/signals.py:
```diff
@@ -0,0 +1,81 @@
+    ParentheticalGroup,
+)
+
+pa_fields_mapping = {
```
Some documentation of this dict would be really helpful. Maybe a comment first, explaining what the keys/values are, and some comments within it, explaining anything else that's interesting?

albertisfu replied:
Sure, I've added the documentation and some comments within the dict.

mlissner commented on cl/search/signals.py:
```diff
@@ -0,0 +1,81 @@
+                "docket_id": ("docket_id", "raw"),
+                "judges": ("judge", "raw"),
+                "nature_of_suit": ("suitNature", "raw"),
+                "precedential_status": (
+                    "status",
+                    "display",
+                ),  # This should be returned as display()
```
Could you get rid of all the `raw` values by making this `get_precedential_status_display`, and having code that tries to invoke the attribute, and failing that, tries to `get` it?

If you did that, this whole thing would get simpler. For example, you could make each thing a `dict[str, str]` instead of `dict[str, tuple[str, str]]`:

```
"opinion__cluster": {
    "case_name": "caseName",
    "citation_count":  "citeCount",
    "date_filed": "dateFiled",
    "slug": "opinion_cluster_slug",
    "docket_id": "docket_id",
    "judges": "judge",
    "nature_of_suit": "suitNature",
    "get_precedential_status_display": "status",
}
```

I also worry a bit about how this reproduces so much of the parenthetical elastic search document object, but perhaps that's unavoidable?

mlissner commented on cl/search/tests.py:
```diff
@@ -2537,6 +2537,235 @@ async def test_pa_search_form_search_and_filtering(self) -> None:
             msg="'Riley' should come before 'Peck' and before 'Smith' when order_by asc.",
         )
 
+    def test_keep_in_sync_related_pa_objects(self) -> None:
```
This is a long test with a lot of IO. Would it be possible to make it into at least two or maybe three or more? When django runs tests in parallel, long tests can tie up a CPU and slow things down while shorter ones can be parallelized.

It's OK if breaking this up doesn't make sense, but if there's an easy and intuitive way to do so, we probably should.

albertisfu replied:
Indeed, it's a good idea to split the ES tests, including dividing them into different files. This is something Eduardo and I have been discussing in order to make search tests more maintainable, so we can work on different modules/documents simultaneously.

So, I believe this is a great opportunity to start splitting tests, since we'll update the ES open PRs with this code once it's merged.

The new structure for search tests is:

- search
  - tests
    - __init__.py
    - tests.py
    - tests_es_parenthetical.py
    
`tests.py` is the same file we had only `ElasticSearchTest` class and its dependencies were removed.

`tests_es_parenthetical.py`, include `ElasticSearchTest` (renamed to ParentheticalESTest) and created a new class `ParentheticalESSignalProcessorTest` for the `ESSignalProcessorTest` tests  split into 4 test methods.

There were some complications doing this ES test split.

- When running tests in `parallel` test methods for different class in the same file can run in parallel.
- The problem of running test methods concurrently from test classes that populate the same elasticsearch index is that it creates conflicts on the shared index, including documents mismatch that lead tests to fail.
- This problem can't be prevented using `lockfile = __file__` since it only works to avoid running test classes concurrently when they're in different files.
- In order to solve the problem is required to ES isolate indices on every test class. I tried different methods but the most simpler is to use a different index name on each test class, so documents for the same type are populated in different indices, this is a single line that needs to be added before populating the ES index:
`parenthetical_group_index._name = cls.__name__.lower()`, in this example `parenthetical_group_index` is the index object on `documents.py`
ChatGPT
Here's the corrected version:

There were some complications doing this test split:

- When running tests in `parallel`, test methods from different classes in the same file can run concurrently.
- The issue with running test methods concurrently from classes that populate the same Elasticsearch index is that it creates conflicts on the shared index. This includes document mismatches that cause tests to fail.
- This problem cannot be prevented using `lockfile = __file__` as it only works to avoid running test classes concurrently when they're in different files.
- To solve this problem, it's necessary to isolate the ES indices for every test class. I tried different methods, but the simplest one is to use a different index name for each test class. That way, documents of the same type are populated in different indices. This requires a single line to be added before populating the ES index: parenthetical_group_index._name = cls.__name__.lower(). In this example, `parenthetical_group_index` is the index object in `documents.py`.

mlissner replied:
ChatGPT has entered the chatroom? What's going on here? ")

albertisfu replied:
haha sorry I use it to correct my grammar 😅

mlissner replied:
Wild world. Anyway, sounds good.

mlissner reviewed (changes requested):
This looks good to me, thank you. 

A few comments. I assume this is safe to merge?

albertisfu commented:
Thanks! I've applied the requested changes. I've made some related comments above.

In addition, I introduced the following changes:

- I added the setting `ELASTICSEARCH_DSL_AUTO_REFRESH = False`. It's recommended to set this to False as it's a resource-intensive operation. It's better to wait for ES to refresh indices every ~1 second.
- For testing, this setting is overridden to `True`, so it's not required to call `ParentheticalGroup._index.refresh()` every time a new document is added/updated. This simplifies tests.
- Include the main_model main to the ESSignalProcessor signals `dispatch_uid` so using it on more document types, the `dispatch_uid` is unique.


> I assume this is safe to merge?

Yeah since the cluster is now running and `ELASTICSEARCH_DISABLED` is set to `False` in prod, this is safe to merge, we'll expect `ParentheticalGroup` will continue being indexed and their related objects also will trigger updates.

mlissner commented on cl/search/tests/tests_es_parenthetical.py:
```diff
@@ -0,0 +1,737 @@
+    @classmethod
+    def setUpTestData(cls):
+        # Create a unique index for this test case
+        parenthetical_group_index._name = cls.__name__.lower()
+        cls.rebuild_index("search.ParentheticalGroup")
```
This is neat, but shouldn't it go on the ESIndexTestCase instead?

albertisfu replied:
Yeah, I thought about it, however the index to use in test classes will depend on the test on the class.

For example:

If running test related to OA the index name will be: `oral_arguments_index`

So I think it's not possible to automatically set the index name in `ESIndexTestCase` since this class is expected to be used on different documents. So it should be manually set in Test classes as required.

mlissner replied:
Wouldn't it be fine in that case if it was dynamically named?

albertisfu replied:
Correct, I changed the method to generate a unique name for each index on tests. I created a new file `es_indices` in order to register all our indices. 

Also we should add each index to:
`es_indices_registered = [parenthetical_group_index]`

So that in `ESIndexTestCase` we set an unique name (based on the test class) for each ES index registered in `es_indices_registered` ensuring that all the indices get a unique name for testing purposes.

mlissner reviewed (commented):
Looks good. One comment remaining.
